src/reflect/scala/reflect/internal/util/FreshNameCreator.scala

```diff
@@ -20,8 +20,13 @@ class FreshNameCreator(creatorPrefix: String = "") {
    */
   def newName(prefix: String): String = {
     val safePrefix = NameTransformer.encode(prefix)
-    counters.putIfAbsent(safePrefix, new AtomicLong(0))
-    val idx = counters.get(safePrefix).incrementAndGet()
+    var counter = counters.get(safePrefix)
+    if (counter eq null) {
+      val fresh = new AtomicLong(0)
+      val raced = counters.putIfAbsent(safePrefix, fresh)
+      counter = if (raced eq null) fresh else raced
+    }
+    val idx = counter.incrementAndGet()
```

Comment:
I have been playing with jmh to get a view of the different options here, together with a POC allocation profiler

I cant submit all of the jmh benchmark for this code as the current benchmark sub-project doesn't depend on reflect or the compiler, and I cant see how to make it do so. 
I will follow up directly with @retronym on this

It seems to me that CA_SB is probably the best option, as it is  the best compromise of readable and performant, and I will update the PR to this option, but happy to hear other opinions

in summary roughly 60% reduction in allocation and CPU usage. Not that hot a method though so probably not noticable for most compiles

Details below
```

[info] Benchmark                                       (param)  Mode  Cnt    Score    Error  Units
[info] FreshNameCreatorBenchmark.orig                        0  avgt   20  147.123 ▒ 10.832  ns/op
[info] FreshNameCreatorBenchmark.orig:allocated              0  avgt   20  319.987 ▒  0.001   B/op
[info] FreshNameCreatorBenchmark.orig_CA                     0  avgt   20  140.252 ▒  0.485  ns/op
[info] FreshNameCreatorBenchmark.orig_CA:allocated           0  avgt   20  311.987 ▒  0.001   B/op
[info] FreshNameCreatorBenchmark.orig_CA_SB                  0  avgt   20   55.228 ▒  0.278  ns/op
[info] FreshNameCreatorBenchmark.orig_CA_SB:allocated        0  avgt   20  131.617 ▒  0.040   B/op
[info] FreshNameCreatorBenchmark.orig_PR                     0  avgt   20  135.632 ▒  0.894  ns/op
[info] FreshNameCreatorBenchmark.orig_PR:allocated           0  avgt   20  311.988 ▒  0.001   B/op
[info] FreshNameCreatorBenchmark.orig_PR_SB                  0  avgt   20   52.463 ▒  0.118  ns/op
[info] FreshNameCreatorBenchmark.orig_PR_SB:allocated        0  avgt   20  131.821 ▒  0.023   B/op

```
the test results correspond to the code options below, 
roughly 
orig - the current code in 2.12
PR - the code i originally proposed in the PR
CA - the code Jason suggested using computeIfAbsent

_SB suffix builds the string directly, rather then using String interpolation

```

  def newNameOrig(prefix: String): String = {
    val safePrefix = NameTransformer.encode(prefix)
    counters.putIfAbsent(safePrefix, new AtomicLong(0))
    val idx = counters.get(safePrefix).incrementAndGet()
    s"$creatorPrefix$safePrefix$idx"
  }
  def newNamePR(prefix: String): String = {
    val safePrefix = NameTransformer.encode(prefix)
    var counter = counters.get(safePrefix)
    if (counter eq null) {
      val fresh = new AtomicLong(0)
      val raced = counters.putIfAbsent(safePrefix, fresh)
      counter = if (raced eq null) fresh else raced
    }
    val idx = counter.incrementAndGet()
    s"$creatorPrefix$safePrefix$idx"
  }
  def newNamePR_SB(prefix: String): String = {
    val safePrefix = NameTransformer.encode(prefix)
    var counter = counters.get(safePrefix)
    if (counter eq null) {
      val fresh = new AtomicLong(0)
      val raced = counters.putIfAbsent(safePrefix, fresh)
      counter = if (raced eq null) fresh else raced
    }
    val idx = counter.incrementAndGet()
    creatorPrefix + safePrefix + idx
  }
  def newNameCA(prefix: String): String = {
    val safePrefix = NameTransformer.encode(prefix)
    val counter = counters.computeIfAbsent(safePrefix, (s: String) => new AtomicLong(0))
    val idx = counter.incrementAndGet()
    s"$creatorPrefix$safePrefix$idx"
  }
  def newNameCA_SB(prefix: String): String = {
    val safePrefix = NameTransformer.encode(prefix)
    val counter = counters.computeIfAbsent(safePrefix, (s: String) => new AtomicLong(0))
    val idx = counter.incrementAndGet()
    creatorPrefix + safePrefix + idx
  }

```